include/xrpl/protocol/STEitherAmount.h

```diff
@@ -0,0 +1,313 @@
+
+// Currency or MPT issuance ID
+template <typename T>
+concept ValidAssetType =
+    std::is_same_v<T, Currency> || std::is_same_v<T, uint192>;
+
+template <typename T>
+concept EitherAmountType = std::is_same_v<T, STEitherAmount> ||
+    std::is_same_v<T, std::optional<STEitherAmount>>;
+
+class STEitherAmount : public STBase, public CountedObject<STEitherAmount>
+{
+private:
+    std::variant<STAmount, STMPTAmount> amount_;
+
+public:
+    using value_type = STEitherAmount;
+    STEitherAmount() = default;
+    STEitherAmount(SerialIter& sit, SField const& name);
+    STEitherAmount(XRPAmount const& amount);
+    STEitherAmount(STAmount const& amount);
+    STEitherAmount(SField const& name, STAmount const& amount = STAmount{});
+    STEitherAmount(SField const& name, STMPTAmount const& amount);
+    STEitherAmount(STMPTAmount const& amount);
+
+    STEitherAmount&
+    operator=(STAmount const&);
+    STEitherAmount&
+    operator=(STMPTAmount const&);
+    STEitherAmount&
+    operator=(XRPAmount const&);
+
+    SerializedTypeID
+    getSType() const override;
+
+    std::string
+    getFullText() const override;
+
+    std::string
+    getText() const override;
+
+    Json::Value getJson(JsonOptions) const override;
+
+    void
+    setJson(Json::Value&) const;
+
+    void
+    add(Serializer& s) const override;
+
+    bool
+    isEquivalent(const STBase& t) const override;
+
+    bool
+    isDefault() const override;
+
+    //------------------------------------------------------------------------------
+
+    bool
+    isMPT() const;
+
+    bool
+    isIssue() const;
+
+    STEitherAmount const&
+    value() const;
+
+    std::variant<STAmount, STMPTAmount> const&
+    getValue() const;
+
+    std::variant<STAmount, STMPTAmount>&
+    getValue();
+
+    AccountID const&
+    getIssuer() const;
+
+    bool
+    negative() const;
+
+    bool
+    native() const;
+
+    STEitherAmount
+    zeroed() const;
+
+    int
+    signum() const noexcept;
+
+    template <ValidAmountType T>
+    T const&
+    get() const;
+
+    template <ValidAmountType T>
+    T&
+    get();
+
+private:
+    STBase*
+    copy(std::size_t n, void* buf) const override;
+    STBase*
+    move(std::size_t n, void* buf) override;
+};
+
+template <ValidAmountType T>
+T const&
+STEitherAmount::get() const
+{
+    if (std::holds_alternative<T>(amount_))
+        return std::get<T>(amount_);
+    Throw<std::logic_error>("Invalid STEitherAmount conversion");
+}
+
+template <ValidAmountType T>
+T&
+STEitherAmount::get()
+{
+    if (std::holds_alternative<T>(amount_))
+        return std::get<T>(amount_);
+    Throw<std::logic_error>("Invalid STEitherAmount conversion");
+}
+
+template <ValidAmountType T>
+decltype(auto)
+get(auto&& amount)
+{
+    using TAmnt = std::decay_t<decltype(amount)>;
+    if constexpr (std::is_same_v<TAmnt, STEitherAmount>)
+    {
+        if constexpr (std::is_lvalue_reference_v<decltype(amount)>)
+            return amount.template get<T>();
```

Comment:
would this call the either one of two `get()`?:

```
template <ValidAmountType T>
T const&
STEitherAmount::get() const
{
    if (std::holds_alternative<T>(amount_))
        return std::get<T>(amount_);
    Throw<std::logic_error>("Invalid STEitherAmount conversion");
}

template <ValidAmountType T>
T&
STEitherAmount::get()
{
    if (std::holds_alternative<T>(amount_))
        return std::get<T>(amount_);
    Throw<std::logic_error>("Invalid STEitherAmount conversion");
```
It seems like both of them would throw.

Reply:
And, yes one of them is going to be called and will throw if the requested type doesn't match the type that the variant holds.